document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoDatastore.java

```diff
@@ -1,79 +1,89 @@
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoDatabase;
 import com.typesafe.config.Config;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import org.bson.Document;
 import org.hypertrace.core.documentstore.Collection;
 import org.hypertrace.core.documentstore.Datastore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class MongoDatastore implements Datastore {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDatastore.class);
 
   private static final String DEFAULT_DB_NAME = "default_db";
   private MongoClient client;
+  private MongoDatabase database;
 
   @Override
   public boolean init(Config config) {
-    try {
-      if (config.hasPath("url")) {
-        client = new MongoClient(new MongoClientURI(config.getString("url")));
-      } else {
-        String hostName = config.getString("host");
-        int port = config.getInt("port");
-        client = new MongoClient(hostName, port);
-      }
-    } catch (IllegalArgumentException e) {
-        throw new IllegalArgumentException(
-            String.format("Unable to instantiate MongoClient with config:%s",config), e);
+    ConnectionString connString;
+    if (config.hasPath("url")) {
+      connString = new ConnectionString(config.getString("url"));
+    } else {
+      String hostName = config.getString("host");
+      int port = config.getInt("port");
+      connString = new ConnectionString("mongodb://" + hostName + ":" + port);
     }
+
+    MongoClientSettings settings = MongoClientSettings.builder()
+        .applyConnectionString(connString)
+        .retryWrites(true)
+        .build();
+    client = MongoClients.create(settings);
+
+    database = client.getDatabase(DEFAULT_DB_NAME);
     return true;
   }
 
   @Override
   public Set<String> listCollections() {
     Set<String> collections = new HashSet<>();
-    List<String> databaseNames = client.getDatabaseNames();
-    for (String dbName : databaseNames) {
-      Set<String> collectionNames = client.getDB(dbName).getCollectionNames();
-      for (String collectionName : collectionNames) {
-        collections.add(dbName + "." + collectionName);
-      }
+    for (String collectionName : database.listCollectionNames()) {
+      collections.add(database.getName() + "." + collectionName);
     }
     return collections;
   }
 
   @Override
   public boolean createCollection(String collectionName, Map<String, String> options) {
-    DBObject dbObject = null;
-    DBCollection collection =
-        client.getDB(DEFAULT_DB_NAME).createCollection(collectionName, dbObject);
-
-    return collection != null;
+    try {
+      database.createCollection(collectionName);
+    } catch (MongoCommandException e) {
+      if (!"NamespaceExists".equals(e.getErrorCodeName())) {
+        LOGGER.error("Could not create collection: {}", collectionName, e);
+        return false;
+      }
```

Comment:
I think it's still worth Logging the error for other failures too for debugging purpose and not limited to just NamespaceExists

Reply:
I'm logging it for other cases itself @inespo

Reply:
ah yes, missed the "!"